visualization/app/codeCharta/ui/codeMap/codeMap.label.service.ts

```diff
@@ -0,0 +1,146 @@
+import * as THREE from "three";
+import {node} from "./rendering/node";
+import {AngularColors, renderSettings} from "./rendering/renderSettings";
+import {CameraChangeSubscriber, ThreeOrbitControlsService} from "./threeViewer/threeOrbitControlsService";
+import {PerspectiveCamera, Sprite} from "three";
+import {ThreeCameraService} from "./threeViewer/threeCameraService";
+import {ThreeSceneService} from "./threeViewer/threeSceneService";
+
+interface internalLabel {
+    sprite : THREE.Sprite;
+    line : THREE.Line | null;
+    heightValue : number;
+}
+
+export class CodeMapLabelService implements CameraChangeSubscriber {
+
+    public static SELECTOR = "codeMapLabelService";
+
+    private labels : internalLabel[];
+    private LABEL_WIDTH_DIVISOR: number = 2600; // empirically gathered
+    private LABEL_HEIGHT_DIVISOR: number = 50; // empirically gathered
+
+    constructor(private threeOrbitControlsService: ThreeOrbitControlsService,
+                private threeCameraService: ThreeCameraService,
+                private threeSceneService: ThreeSceneService) {
+
+        this.labels = new Array<internalLabel>();
+        threeOrbitControlsService.subscribe(this);
+    }
+
+    addLabel(node: node, settings: renderSettings) : void {
+        if(node.attributes && node.attributes[settings.heightKey]){
+
+            const x: number = node.x0 - settings.mapSize * 0.5;
+            const y: number = node.z0;
+            const z: number = node.y0 - settings.mapSize * 0.5;
```

Comment:
Does anybody have an idea, why `node.z0` is applied to the variable `y` and `node.y0...` applied to the variable `z`? It has also been like this before (see [labelManager.ts#L25](https://github.com/MaibornWolff/codecharta/blob/master/visualization/app/codeCharta/ui/codeMap/rendering/labelManager.ts#L25))

Reply:
This is actually something we should make clear in code, maybe create a quality item for it. I am not a 100% sure but I think it has to do with the way three.js arranges the axis. When we treemap our data with d3 we naturally get x/y coordinates since we are two-dimensional. Our treemap service adds a third dimension to this treemap which is z. 
As soon we want to use our treemap in three.js we need to map our axis [to the three.js axis which look like this  ](https://www.oreilly.com/library/view/webgl-up-and/9781449326487/httpatomoreillycomsourceoreillyimages1276316.png.jpg) (y is up, not z).

Reply:
A task for a quality day. Check where it happens and extract the switch logic into a method with a meaningful name. This way the code stays self explanatory without further documentation. I will create it.

Reply:
Ahh, sure 👍 thanks!